### PR TITLE
feat: eth_getLogs

### PR DESCRIPTION
**Pull request by greged93**
Provides the `eth_getLogs` endpoint in order to retrieve ethereum logs emitted by the kakarot contract.

Time spent on this PR: 2 days

Resolves: #398 

# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing

# What is the new behavior?
Provides the `eth_getLogs` endpoint with unit testing and integration testing.

# Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: crates/core/tests/client.rs [hunk 1]
```diff
@@ -239,4 +190,85 @@ mod tests {
             balances
         );
     }
+
+    #[tokio::test]
+    async fn test_get_logs() {
```
Comment (ftupas) [inline crates/core/tests/client.rs:195]: I think the test fixture `#[with(TestContext::ERC20)] kakarot_test_env_ctx: KakarotTestEnvironmentContext` can be useful here 🙂

**Review by ftupas — COMMENTED** [review]
Good stuff, really cool pr 💪 Just some minor comments 🙂

File: crates/core/src/models/event_filter.rs [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    use super::*;
+    use crate::mock::constants::KAKAROT_ADDRESS;
+    use crate::mock::mock_starknet::{fixtures, init_mock_client};
+
+    fn assert_eq_event_filter(lhs: EventFilter, rhs: EventFilter) {
+        assert_eq!(lhs.from_block, rhs.from_block);
+        assert_eq!(lhs.to_block, rhs.to_block);
+        assert_eq!(lhs.address, Some(*KAKAROT_ADDRESS));
+        assert_eq!(lhs.keys, rhs.keys);
+    }
```
[the remark this answers is not in the document]
Comment (greged93) [reply crates/core/src/models/event_filter.rs:113]: EventFilter is from starknet-rs so can't do that :/

File: crates/core/src/client/helpers.rs [hunk 3]
```diff
@@ -123,9 +123,21 @@ pub fn raw_kakarot_calldata(kakarot_address: FieldElement, mut calldata: Vec<Fie
...
 
+/// Helper function to split a U256 value into two FieldElements.
+pub fn split_u256_into_field_elements(value: U256) -> [FieldElement; 2] {
+    let low = value & U256::from(U128::MAX);
+    let high = value >> 128;
+    [
+        FieldElement::from_bytes_be(&low.to_be_bytes()).unwrap(), // Safe unwrap <= U128::MAX.
+        FieldElement::from_bytes_be(&high.to_be_bytes()).unwrap(), // Safe unwrap <= U128::MAX.
+    ]
+}
```
[the remark this answers is not in the document]
Comment (greged93) [reply crates/core/src/client/helpers.rs:134]: done!

File: crates/core/src/client/tests/mod.rs [hunk 4]
```diff
@@ -189,3 +189,43 @@ async fn test_gas_price() {
...
+
+    // When
+    let logs = client.get_logs(filter).await.unwrap();
+
+    // Then
+    assert!(logs.is_empty());
+}
+
+#[tokio::test]
+async fn test_get_logs_to_less_than_from() {
```
[the remark this answers is not in the document]
Comment (greged93) [reply crates/core/src/client/tests/mod.rs:214]: done!

File: crates/core/src/models/event_filter.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ftupas) [reply crates/core/src/models/event_filter.rs:113]: Gotcha, this should be fine 👌

File: crates/core/src/test_utils/deploy_helpers.rs [hunk 5]
```diff
@@ -492,6 +493,7 @@ async fn deploy_kakarot(
         *class_hash.get("contract_account").unwrap(),
         *class_hash.get("externally_owned_account").unwrap(),
         *class_hash.get("proxy").unwrap(),
+        *DEPLOY_FEE,
```
Comment (ftupas) [inline crates/core/src/test_utils/deploy_helpers.rs:496]: Thanks for adding this!

File: crates/core/src/test_utils/execution_helpers.rs [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+use crate::models::felt::Felt252Wrapper;
+
+pub async fn execute_tx(env: &KakarotTestEnvironmentContext, contract: &str, selector: &str, args: Vec<U256>) -> H256 {
+    let contract = env.evm_contract(contract);
+    let contract_eth_address = {
+        let address: Felt252Wrapper = contract.addresses.eth_address.into();
+        address.try_into().unwrap()
+    };
+    let client = env.client();
+    let kakarot = env.kakarot();
```
Comment (ftupas) [inline crates/core/src/test_utils/execution_helpers.rs:14]: I believe we can use [`resources_with_contract`](https://github.com/kkrt-labs/kakarot-rpc/blob/main/crates/core/src/test_utils/deploy_helpers.rs#L754C4-L754C4) here as well

```rust
let (client, kakarot, contract, contract_eth_address) = env.resources_with_contract();
```

File: crates/core/src/client/tests/mod.rs [hunk 7]
```diff
@@ -189,3 +189,100 @@ async fn test_gas_price() {
...
+    };
+
+    // When
+    let logs = client.get_logs(filter).await.unwrap();
+
+    // Then
+    assert_eq!(2, logs.len());
+    assert_eq!(
+        Log {
+            address: *ABDEL_ETHEREUM_ADDRESS,
```
Comment (ftupas) [inline crates/core/src/client/tests/mod.rs:254]: Nice 🚀🫡

File: crates/core/src/client/tests/mod.rs [hunk 8]
```diff
@@ -189,3 +189,100 @@ async fn test_gas_price() {
...
+
+    // When
+    let logs = client.get_logs(filter).await.unwrap();
+
+    // Then
+    assert!(logs.is_empty());
+}
+
+#[tokio::test]
+async fn test_get_logs() {
```
Comment (ftupas) [inline crates/core/src/client/tests/mod.rs:234]: Awesome 👌

**Review by ftupas — DISMISSED** [review]
Just a minor refactor otherwise looks really good! ✨

File: crates/core/src/test_utils/execution_helpers.rs (hunk 6 above)
Comment (greged93) [reply crates/core/src/test_utils/execution_helpers.rs:14]: ah damn you are right !

**Review by ftupas — APPROVED** [review]
Looks good!

**Pull request merged**